### ISSUE TITLE
Improved tag selection flow

### ISSUE THREAD

**Issue by asimonw** [emfoundation/ce100-app#789]
This applies to creating/editing organisations/challenges.

Rather than having a two step process, where the user first adds/edits other content to an org profile or challenge and then clicks next to add tags, the tag selection will be more integrated.

The currently selected (if any) tags are shown in the edit form with a "Add/Edit tags" button below them. Clicking on that button opens up the tag selection screen. On that screen, clicking either "Select" or "Cancel" takes you back to the edit form. Only after clicking "Save" on that form are the changes to the tags saved. 

Linked to issues:
+ [x]  #815 
+ [ ]  #801 
+ [x]  #808 
+ [ ]  #790

**Comment by fannyBkane** [comment]
[
![screen shot 2017-05-09 at 17 18 28](https://cloud.githubusercontent.com/assets/20577339/25861076/94393676-34db-11e7-9e11-57555a3d7439.png)
](url)

**Comment by fannyBkane** [comment]
Form/mobile: https://zpl.io/W62Dw
Modal/mobile: https://zpl.io/222375

**Comment by SimonLab** [comment]
Main steps:
- When the new tags are saved we need to save the tags state on the frontend with cookiee/seesion/localstorage
- redirect to the edit page and display the new selected tags on the view based on the cookie
- on save we need first to save the tags then to save the organisation details
- on cancel we don't save anything and redirect to the previous page

We need to choose the right structure for saving tags in cookies and refactor the saving function on the backend.

**Comment by asimonw** [comment]
@SimonLab Are you trying to avoid rendering the tag selection screen in the browser? I think that won't work if it's a modal on desktop.

**Comment by SimonLab** [comment]
sorry @asimonw I'm not sure I understand your previous comment and my previous comment wasn't clear also. We had a bit more thought and will had more detail of the implementation. The idea is to reuse as much as possible of the existing code and logic to avoid recreating everything from scratch

- [ ] When a user edit or create a new organisation/challenge we are storing in localstorage the values of the selected tags (ie save on the browser the initial state of the challenge/org)
- [ ] we merge the edit org/challenge and add tags pages. The list of tags will be hidden first then on click "edit tags" the list will be displayed.
- [ ] on save tags we are reusing our current logic and save the tags in the database
- [ ] on save org/chall nothing more has to be done as the tags are alredy saved in the database
- [ ] on cancel we are getting the initial state or the tags from the localstorage then update the database with these tags (for a new org/chall the initial tags state is empty)

We have created new issues for these points, this allow us to add estimation on each of them and add more detail and explanation during our implementation process

**Comment by asimonw** [comment]
@SimonLab This doesn't fully correspond to the requirements, though. The tag selection screen doesn't have a "Save" option, only "Select" or "Cancel". When clicking "Select" the data should not yet be saved to the database. The data should only be persisted once the user clicks "Save" on the org/challenge form. Note that the user can still cancel the whole action even after "selecting tags". In other words, items 3 and 4 do not correspond to the outlined user flow.

My earlier comment was about rendering the tag selection view in the client, but it sounds like you'll render it on the server and simply hide it when loading the form. This will increase initial page load, but would make opening up the tag selection faster. That could be a good trade-off for now, we'll test how it affects the user experience.

**Comment by asimonw** [comment]
Also, you might not need localStorage. Just saving the state as a JS object in memory would be fine, no? What would be the reason for keeping that data around once the user has submitted the form?

**Comment by SimonLab** [comment]
@asimonw thanks for your comments. you have some good points and your logic was the one I had in mind at the begining.

We are saving the tags on "select" as at the moment it is the easier and quicker to implement the flow. The final logic will however be the one describe on your top comment. If the user cancel the whole action we will reset the tags in the database to the initial state.
This way we minimise the code we need to rewrite for the front end and we reuse the code of the backend.

Yes rendering the page with all the tags will be a bit longer(it might be negligible I think) but I think it might be a bit more fluid instead of sending some ajax requests to get the list of tags.

You are right about the localstorage, we might not need it. However using only javascript to store the state will create more global variables and more code on the frontend which will make the page load a bit longer (but maybe it's negligible). Also I think localstorage is a bit safer and less accessible to the users than js. We will also clear the localstorage data as soon as the org/challenge are saved or canceled

To recap I think our solution will work, I understand it can feels a bit strange to save the tags on each select, but again the idea is to reduce the number of javascript code on the front end and to reuse and adapt the existing code. But agree to explore a more javascript oriented solution and to compare them :+1:

**Comment by SimonLab** [comment]
@asimonw saving tags on select will be an issue if the user change the url manually, this means that we have saved the tags in the database even if the user hasn't click on the "save" button. So you are right and we are going to refactor to only save when the user click on the save. Thanks for your comments :+1:

**Comment by SimonLab** [comment]
next steps for implementing the javascript state (exemple with challenge edit page):

on edit challenge load page event (ex: challenges/3/edit)
- [ ] save in js variable the list of the current selected categories and tags

On "edit tags" click event
- [ ] select the categories based on the current state of the categories (uncheck to reset + check the  selected categories)
- [ ] select the tags based on the current state of the tags
- [ ] open modal view (change the css "display" property to "block" as the tags are already loaded on the page)

On "select tags" click event
- [ ] Check that the number of selected tags < 10 - if not display error message on modal
- [ ] save the selected tags on the current state tags variable
- [ ] save the selected categories on the current state categories variable
- [ ] load name of tags on the main page:
  ![image](https://cloud.githubusercontent.com/assets/6057298/26151050/0d0c6212-3af9-11e7-9ab9-7153f7520d93.png)
- load hidden inputs to represent the selected tags on the form

On "cancel" modal event
- [ ] close the modal

On "Save and Publish" click event
- [ ] submit form with the new payload which contains the list of tags id
- [ ] update Joi which validate the new payload
- [ ] on the backend call the save tag function in the handler

**Comment by SimonLab** [comment]
@asimonw from your top comment "This applies to creating/editing organisations/challenges."
Do we also want the same modal tag view and flow for the insights?

**Comment by asimonw** [comment]
@SimonLab I would definitely want to consider it. The flow shouldn't change in the explore view (after selecting a tag, you see the filtered list of orgs, challenges or insights), but on desktop we could use a modal to keep the look and feel of tag selection consistent throughout the app.

On mobile, this should have no effect on the ux. Only on desktop, the tag selection screen could be a modal rather than full screen. How much time would it take you to do that once the changes have been made for the forms? I assume that you would be able to reuse a lot of the code?

**Comment by sohilpandya** [comment]
@fannyBkane the screenshot that you've uploaded above for `edit-profile-mobile-copy` is not in zeppelin, I wanted to update the `edit tags` button but have no reference to it. 

It would be great to get access to that file as well. 

Thanks 😊

**Comment by fannyBkane** [comment]
Hey @sohilpandya  my bad. Here are the links:

Form/Edit profile/mobile: https://zpl.io/W62Dw
Form/Edit profile/mobile: https://zpl.io/222375

**Comment by asimonw** [comment]
@SimonLab Works beautifully thanks! Just one redirect issue for challenges.

If you follow the following steps:
* create or edit a challenge and click save
* you end up on the challenge card, which is perfect
* click cancel on the challenge card 
* you go back to the "Edit challenge" view. 

It would make more sense to direct the user to the org profile at this stage.

The redirect is a bit subtle, because you can also end up on a challenge card through the explore screen. So, could you make sure the following happens: 
* If the previous page was Explore: cancel challenge card -> back to Explore (already works like that)
* If the previous page was org profile or edit challenge -> to org profile

Let me know if that doesn't make sense.

**Comment by asimonw** [comment]
@SimonLab Apologies, this should be addressed in #790 where the issue was previously raised. I've copied my comment to that issue and will close this one.
